src/api.py

```diff
@@ -295,50 +333,29 @@ def _generate_with_retry() -> Any:
         for output_block in output_blocks:
             output_block.finish_stream()
 
-        usage_reports = self._calculate_usage(messages, output_texts, completion_id)
-
-        return usage_reports
+        return billing_callback.usage(completion_id)
 
     def _reassemble_function_call(self, function_call_chunks: List[dict]) -> dict:
         result = {}
         for chunk in function_call_chunks:
-            for key in chunk.keys():
+            model = chunk.model_dump()
+            for key in model.keys():
                 if key not in result:
                     result[key] = ""
-                result[key] += chunk[key]
+                if model[key]:
+                    result[key] += model[key]
         return result
 
-    def _calculate_usage(
-        self,
-        messages: List[Dict[str, str]],
-        output_texts: List[str],
-        completion_id: str,
-    ) -> [UsageReport]:
-        # for token usage tracking, we need to include not just the token usage, but also completion id
-        # that will allow proper usage aggregation for n > 1 cases
-        encoding = tiktoken.encoding_for_model(self.config.model)
-        output_tokens = sum([len(encoding.encode(text)) for text in output_texts])
-        prompt_tokens = sum(
-            [
-                len(encoding.encode(message.get("content", "") or ""))
-                for message in messages
-            ]
-        )
-        usage_reports = [
-            UsageReport(
-                operation_type=OperationType.RUN,
-                operation_unit=OperationUnit.PROMPT_TOKENS,
-                operation_amount=prompt_tokens,
-                audit_id=completion_id,
-            ),
+    def _calculate_usage_from_completion(self, completion_response, completion_id) -> [UsageReport]:
+        cost = completion_cost(completion_response=completion_response)
+        return [
             UsageReport(
                 operation_type=OperationType.RUN,
-                operation_unit=OperationUnit.SAMPLED_TOKENS,
-                operation_amount=output_tokens,
-                audit_id=completion_id,
-            ),
+                operation_unit=OperationUnit.UNITS,
+                operation_amount=int(cost * 10_000_000_000_000),
```

Comment:
I think I understand this from previous discussions during a meeting, but a comment or two here would be super :)

Reply:
Oh whoops, that shouldn't be appearing in the code twice, one of them was dead that I was using for reference to the new billing callback method.  Yanked that but broke out a constant anyway and doc'd it